<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Orthology inference dNdS estimation of orthologous genes using orthologr • orthologr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Orthology inference dNdS estimation of orthologous genes using orthologr">
<meta property="og:description" content="orthologr">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">Orthology inference dNdS estimation of orthologous genes using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/drostlab/orthologr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Orthology inference dNdS estimation of
orthologous genes using orthologr</h1>
            
            <h4 data-toc-skip class="date">2023-10-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/drostlab/orthologr/blob/HEAD/vignettes/dNdS_estimation.Rmd" class="external-link"><code>vignettes/dNdS_estimation.Rmd</code></a></small>
      <div class="hidden name"><code>dNdS_estimation.Rmd</code></div>

    </div>

    
    
<p>The dN/dS ratio quantifies the mode and strength of selection acting
on a pair of orthologous genes. This selection pressure can be
quantified by comparing synonymous substitution rates (dS) that are
assumed to be neutral with nonsynonymous substitution rates (dN), which
are exposed to selection as they change the amino acid composition of a
protein (<a href="http://mbe.oxfordjournals.org/content/31/1/212" class="external-link">Mugal
et al., 2013</a>).</p>
<p>The <code>orthologr</code> package provides a function named
<code><a href="../reference/dNdS.html">dNdS()</a></code> to perform dNdS estimation on pairs of orthologous
genes. The <code><a href="../reference/dNdS.html">dNdS()</a></code> function takes the CDS files of two
organisms of interest (<code>query_file</code> and
<code>subject_file</code>) and computes the dNdS estimation values for
orthologous gene pairs between these organisms.</p>
<p><strong>Note:</strong> the following dNdS estimation methods are
based on KaKs_Calculator 2:</p>
<ul>
<li><p>“NG”: Nei, M. and Gojobori, T. (1986)</p></li>
<li><p>“LWL”: Li, W.H., et al. (1985)</p></li>
<li><p>“LPB”: Li, W.H. (1993) and Pamilo, P. and Bianchi, N.O.
(1993)</p></li>
<li><p>“MLWL” (Modified LWL), MLPB (Modified LPB): Tzeng, Y.H., et
al. (2004)</p></li>
<li><p>“YN”: Yang, Z. and Nielsen, R. (2000)</p></li>
<li><p>“MYN” (Modified YN): Zhang, Z., et al. (2006)</p></li>
<li><p>“GMYN”: Wang, D.P., et al. Biology Direct. (2009)</p></li>
<li><p>“GY”: Goldman, N. and Yang, Z. (1994)</p></li>
<li><p>“MS”: (Model Selection): based on a set of candidate models,
Posada, D. (2003)</p></li>
<li><p>“MA” (Model Averaging): based on a set of candidate models,
Posada, D. (2003)</p></li>
<li><p>“ALL”: All models toghether</p></li>
</ul>
<p>It is assumed that when you choose one of these dNdS estimation
methods you have <a href="https://sourceforge.net/projects/kakscalculator2/" class="external-link">KaKs_Calculator
2</a> installed on your machine and it can be executed from the default
execution <code>PATH</code>.</p>
<p>The following pipeline resembles an example dNdS estimation
procedure:</p>
<ol style="list-style-type: decimal">
<li><p>Orthology Inference: e.g. BLAST reciprocal best hit
(RBH)</p></li>
<li><p>Pairwise sequence alignment: e.g. clustalw for pairwise amino
acid sequence alignments</p></li>
<li><p>Codon Alignment: e.g. pal2nal program</p></li>
<li><p>dNdS estimation: e.g. <a href="http://mbe.oxfordjournals.org/content/17/1/32.short" class="external-link">Yang, Z. and
Nielsen, R. (2000)</a> (YN)</p></li>
</ol>
<p><strong>Note:</strong> it is assumed that when using
<code><a href="../reference/dNdS.html">dNdS()</a></code> all corresponding multiple sequence alignment
programs you want to use are already installed on your machine and are
executable via either the default execution PATH or you specifically
define the location of the executable program via the
<code>aa_aln_path</code> or <code>blast_path</code> argument that can be
passed to <code><a href="../reference/dNdS.html">dNdS()</a></code>. See the <a href="https://github.com/HajkD/orthologr/blob/master/vignettes/sequence_alignments.Rmd" class="external-link">Sequence
Alignments vignette</a> for details.</p>
<p>The following example shall illustrate a dNdS estimation process.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drostlab/orthologr" class="external-link">orthologr</a></span><span class="op">)</span></span>
<span><span class="co"># get a dNdS table using:</span></span>
<span><span class="co"># 1) reciprocal best hit for orthology inference (RBH)</span></span>
<span><span class="co"># 2) Needleman-Wunsch for pairwise amino acid alignments</span></span>
<span><span class="co"># 3) pal2nal for codon alignments</span></span>
<span><span class="co"># 4) Comeron for dNdS estimation</span></span>
<span><span class="co"># 5) single core processing 'comp_cores = 1'</span></span>
<span><span class="fu"><a href="../reference/dNdS.html">dNdS</a></span><span class="op">(</span>query_file      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_thal_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span>
<span>     subject_file    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_lyra_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span>
<span>     ortho_detection <span class="op">=</span> <span class="st">"RBH"</span>, </span>
<span>     aa_aln_type     <span class="op">=</span> <span class="st">"pairwise"</span>,</span>
<span>     aa_aln_tool     <span class="op">=</span> <span class="st">"NW"</span>, </span>
<span>     codon_aln_tool  <span class="op">=</span> <span class="st">"pal2nal"</span>, </span>
<span>     dnds_est.method <span class="op">=</span> <span class="st">"Comeron"</span>, </span>
<span>     comp_cores      <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code>  A tibble: 20 x 15
   query_id subject_id      dN    dS   dNdS perc_identity alig_length
   &lt;chr&gt;    &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;         &lt;dbl&gt;       &lt;dbl&gt;
 1 AT1G010... 333554|PA... 0.106   0.254 0.420           74.0         469
 2 AT1G010... 470181|PA... 0.0402  0.104 0.388           91.1         246
 3 AT1G010... 470180|PA... 0.0150  0.126 0.118           95.5         359
 4 AT1G010... 333551|PA... 0.0135  0.116 0.116           92.0        1970
 5 AT1G010... 909874|PA... 0       0.175 0              100           213
 6 AT1G010... 470177|PA... 0.0449  0.113 0.397           89.5         648
 7 AT1G010... 918864|PA... 0.0183  0.106 0.173           95.1         366
 8 AT1G010... 909871|PA... 0.0340  0.106 0.322           90.3         300
 9 AT1G010... 470171|PA... 0.00910 0.218 0.0417          96.8         434
10 AT1G011... 333544|PA... 0.0325  0.122 0.266           93.6         528
11 AT1G011... 918858|PA... 0.00307 0.133 0.0232          99.2         529
12 AT1G011... 470161|PA... 0.00567 0.131 0.0432          98.5         453
13 AT1G011... 918855|PA... 0.13    0.203 0.641           72.6         285
14 AT1G011... 918854|PA... 0.105   0.280 0.373           84.9         179
15 AT1G011... 311317|PA... 0       0.306 0               85.6          97
16 AT1G011... 909860|PA... 0.0297  0.176 0.168           92.6         310
17 AT1G011... 311315|PA... 0.0287  0.162 0.177           94.2         533
18 AT1G012... 470156|PA... 0.0190  0.168 0.114           95.8         238
19 AT1G012... 311313|PA... 0.0207  0.154 0.134           95.3         107
20 AT1G012... 470155|PA... 0.0157  0.153 0.102           96.7        1056
 ... with 8 more variables: mismatches &lt;dbl&gt;, gap_openings &lt;dbl&gt;,
   q_start &lt;dbl&gt;, q_end &lt;dbl&gt;, s_start &lt;dbl&gt;, s_end &lt;dbl&gt;, evalue &lt;dbl&gt;,
   bit_score &lt;dbl&gt;</code></pre>
<p>Some outputs include <code>NA</code> values. To filter for
<code>NA</code> values or a specific <code>dnds.threshold</code>, you
can use the <code><a href="../reference/filter_dNdS.html">filter_dNdS()</a></code> function. The
<code><a href="../reference/filter_dNdS.html">filter_dNdS()</a></code> function takes the output data.table returned
by <code><a href="../reference/dNdS.html">dNdS()</a></code> and filters the output by the following
criteria:</p>
<ol style="list-style-type: decimal">
<li><p>all dN values having an NA value are omitted</p></li>
<li><p>all dS values having an NA value are omitted</p></li>
<li><p>all dNdS values &gt;= the specified <code>dnds.threshold</code>
are omitted</p></li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drostlab/orthologr" class="external-link">orthologr</a></span><span class="op">)</span></span>
<span><span class="co"># get dNdS estimated for orthologous genes between A. thaliana and A. lyrata</span></span>
<span><span class="va">Ath_Aly_dnds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dNdS.html">dNdS</a></span><span class="op">(</span>query_file      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_thal_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span>
<span>     subject_file    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_lyra_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span>
<span>     ortho_detection <span class="op">=</span> <span class="st">"RBH"</span>, </span>
<span>     aa_aln_type     <span class="op">=</span> <span class="st">"pairwise"</span>,</span>
<span>     aa_aln_tool     <span class="op">=</span> <span class="st">"NW"</span>, </span>
<span>     codon_aln_tool  <span class="op">=</span> <span class="st">"pal2nal"</span>, </span>
<span>     dnds_est.method <span class="op">=</span> <span class="st">"Comeron"</span>, </span>
<span>     comp_cores      <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># filter for:</span></span>
<span><span class="co"># 1) all dN values having an NA value are omitted</span></span>
<span><span class="co"># 2) all dS values having an NA value are omitted</span></span>
<span><span class="co"># 3) all dNdS values &gt;= 2 are omitted</span></span>
<span><span class="fu"><a href="../reference/filter_dNdS.html">filter_dNdS</a></span><span class="op">(</span><span class="va">Ath_Aly_dnds</span>, dnds.threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code>Filtering out NA values in dN or dS and all values with dNdS &gt; 2 ...
Initial input contains 20 rows.
Filtering done. New output table contains 20 rows.
 A tibble: 20 x 15
   query_id subject_id      dN    dS   dNdS perc_identity alig_length
   &lt;chr&gt;    &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;         &lt;dbl&gt;       &lt;dbl&gt;
 1 AT1G010... 333554|PA... 0.106   0.254 0.420           74.0         469
 2 AT1G010... 470181|PA... 0.0402  0.104 0.388           91.1         246
 3 AT1G010... 470180|PA... 0.0150  0.126 0.118           95.5         359
 4 AT1G010... 333551|PA... 0.0135  0.116 0.116           92.0        1970
 5 AT1G010... 909874|PA... 0       0.175 0              100           213
 6 AT1G010... 470177|PA... 0.0449  0.113 0.397           89.5         648
 7 AT1G010... 918864|PA... 0.0183  0.106 0.173           95.1         366
 8 AT1G010... 909871|PA... 0.0340  0.106 0.322           90.3         300
 9 AT1G010... 470171|PA... 0.00910 0.218 0.0417          96.8         434
10 AT1G011... 333544|PA... 0.0325  0.122 0.266           93.6         528
11 AT1G011... 918858|PA... 0.00307 0.133 0.0232          99.2         529
12 AT1G011... 470161|PA... 0.00567 0.131 0.0432          98.5         453
13 AT1G011... 918855|PA... 0.13    0.203 0.641           72.6         285
14 AT1G011... 918854|PA... 0.105   0.280 0.373           84.9         179
15 AT1G011... 311317|PA... 0       0.306 0               85.6          97
16 AT1G011... 909860|PA... 0.0297  0.176 0.168           92.6         310
17 AT1G011... 311315|PA... 0.0287  0.162 0.177           94.2         533
18 AT1G012... 470156|PA... 0.0190  0.168 0.114           95.8         238
19 AT1G012... 311313|PA... 0.0207  0.154 0.134           95.3         107
20 AT1G012... 470155|PA... 0.0157  0.153 0.102           96.7        1056
 ... with 8 more variables: mismatches &lt;dbl&gt;, gap_openings &lt;dbl&gt;,
   q_start &lt;dbl&gt;, q_end &lt;dbl&gt;, s_start &lt;dbl&gt;, s_end &lt;dbl&gt;, evalue &lt;dbl&gt;,
   bit_score &lt;dbl&gt;</code></pre>
<p>The <code><a href="../reference/dNdS.html">dNdS()</a></code> function can be used choosing the following
options:</p>
<ul>
<li>
<code>ortho_detection</code> : <code>RBH</code> (BLAST best
reciprocal hit), <code>BH</code> (BLAST best reciprocal hit),
<code>PO</code> (ProteinOrtho), and <code>OrthoMCL</code>
(OrthoMCL)</li>
<li>
<code>aa_aln_type</code> : <code>multiple</code> or
<code>pairwise</code>
</li>
<li>
<code>aa_aln_tool</code> : <code>clustalw</code>,
<code>t_coffee</code>, <code>muscle</code>, <code>clustalo</code>,
<code>mafft</code>, and <code>NW</code> (in case
<code>aa_aln_type = "pairwise"</code>)</li>
<li>
<code>codon_aln_tool</code> : <code>pal2nal</code>
</li>
<li>
<code>dnds_est.method</code> : <code>Li</code>,
<code>Comeron</code>, <code>NG</code>, <code>LWL</code>,
<code>LPB</code>, <code>MLWL</code>, <code>YN</code>, and
<code>MYN</code>
</li>
</ul>
<p>Please see <code><a href="../reference/dNdS.html">?dNdS</a></code> for details.</p>
<p>In case your BLAST program, or multiple alignment program can not be
executed from the default execution <code>PATH</code> you can specify
the <code>aa_aln_path</code> or <code>blast_path</code> arguments.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drostlab/orthologr" class="external-link">orthologr</a></span><span class="op">)</span></span>
<span><span class="co"># using the `aa_aln_path` or `blast_path` arguments</span></span>
<span><span class="fu"><a href="../reference/dNdS.html">dNdS</a></span><span class="op">(</span> query_file      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_thal_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span>
<span>      subject_file    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_lyra_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span>
<span>      ortho_detection <span class="op">=</span> <span class="st">"RBH"</span>,</span>
<span>      blast_path      <span class="op">=</span> <span class="st">"here/path/to/blastp"</span>,</span>
<span>      aa_aln_type     <span class="op">=</span> <span class="st">"multiple"</span>, </span>
<span>      aa_aln_tool     <span class="op">=</span> <span class="st">"clustalw"</span>, </span>
<span>      aa_aln_path     <span class="op">=</span> <span class="st">"here/path/to/clustalw"</span>,</span>
<span>      codon_aln_tool  <span class="op">=</span> <span class="st">"pal2nal"</span>, </span>
<span>      dnds_est.method <span class="op">=</span> <span class="st">"Comeron"</span>, </span>
<span>      comp_cores      <span class="op">=</span> <span class="fl">1</span>, </span>
<span>      clean_folders   <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="advanced-options">Advanced options<a class="anchor" aria-label="anchor" href="#advanced-options"></a>
</h2>
<p>Additional arguments can be passed to <code><a href="../reference/dNdS.html">dNdS()</a></code>. This
allows you to use more advanced options of several interface
programs.</p>
<p>To pass additional parameters to the interface programs, you can use
the <code>blast_params</code> and <code>aa_aln_params</code> arguments.
The <code>aa_aln_params</code> argument assumes that when you chose
e.g. <code>aa_aln_tool = "mafft"</code> you will pass the corresponding
additional parameters in MAFFT notation.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drostlab/orthologr" class="external-link">orthologr</a></span><span class="op">)</span></span>
<span><span class="co"># get dNdS estimated for orthologous genes between A. thaliana and A. lyrata</span></span>
<span><span class="co"># using additional parameters:</span></span>
<span></span>
<span><span class="co"># get a dNdS table using:</span></span>
<span><span class="co"># 1) reciprocal best hit for orthology inference (RBH)</span></span>
<span><span class="co"># 2) multiple amino acid alignments using MAFFT</span></span>
<span><span class="co"># 3) pal2nal for codon alignments</span></span>
<span><span class="co"># 4) Comeron (1995) for dNdS estimation</span></span>
<span><span class="co"># 5) single core processing 'comp_cores = 1'</span></span>
<span><span class="va">Ath_Aly_dnds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dNdS.html">dNdS</a></span><span class="op">(</span> query_file      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_thal_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span>
<span>                      subject_file    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_lyra_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span>
<span>                      ortho_detection <span class="op">=</span> <span class="st">"RBH"</span>,</span>
<span>                      blast_params    <span class="op">=</span> <span class="st">"-matrix BLOSUM80"</span>,</span>
<span>                      aa_aln_type     <span class="op">=</span> <span class="st">"multiple"</span>,</span>
<span>                      aa_aln_tool     <span class="op">=</span> <span class="st">"mafft"</span>,</span>
<span>                      aa_aln_params   <span class="op">=</span> <span class="st">"--maxiterate 1 --clustalout"</span>,</span>
<span>                      dnds_est.method <span class="op">=</span> <span class="st">"Comeron"</span>,</span>
<span>                      comp_cores      <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                      clean_folders   <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                      quiet           <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># filter for:</span></span>
<span><span class="co"># 1) all dN values having an NA value are omitted</span></span>
<span><span class="co"># 2) all dS values having an NA value are omitted</span></span>
<span><span class="co"># 3) all dNdS values &gt;= 0.1 are omitted</span></span>
<span><span class="fu"><a href="../reference/filter_dNdS.html">filter_dNdS</a></span><span class="op">(</span><span class="va">Ath_Aly_dnds</span>, dnds.threshold <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<pre><code>
     query_id      subject_id       dN     dS    dNdS
1 AT1G01050.1 909874|PACid:16 0.000000 0.1750 0.00000
2 AT1G01090.1 470171|PACid:16 0.009843 0.2150 0.04579
3 AT1G01120.1 918858|PACid:16 0.003072 0.1326 0.02317
4 AT1G01140.3 470161|PACid:16 0.005672 0.1312 0.04324
5 AT1G01170.2 311317|PACid:16 0.008750 0.2827 0.03095
6 AT1G01220.1 470155|PACid:16 0.015210 0.1533 0.09919
</code></pre>
<p>Here <code>blast_params</code> and <code>aa_aln_params</code> take an
character string specifying the parameters that shall be passed to BLAST
and MAFFT. The notation of these parameters must follow the command line
call of the stand alone versions of BLAST and MAFFT:
e.g. <code>blast_params = "blast_params = -matrix BLOSUM80"</code> and
<code>aa_aln_params = "--maxiterate 1 --clustalout"</code>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
