<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Compute a Codon Alignment — codon_aln • orthologr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Compute a Codon Alignment — codon_aln"><meta property="og:description" content="This function takes a protein alignment and the corresponding coding sequences 
as fasta files and computes the corresponding codon alignment based on the models specified in the PAL2NAL program."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">Orthology inference dNdS estimation of orthologous genes using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/drostlab/orthologr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute a Codon Alignment</h1>
    <small class="dont-index">Source: <a href="https://github.com/drostlab/orthologr/blob/HEAD/R/codon_aln.R" class="external-link"><code>R/codon_aln.R</code></a></small>
    <div class="hidden name"><code>codon_aln.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes a protein alignment and the corresponding coding sequences 
as fasta files and computes the corresponding codon alignment based on the models specified in the PAL2NAL program.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">codon_aln</span><span class="op">(</span></span>
<span>  <span class="va">file_aln</span>,</span>
<span>  <span class="va">file_nuc</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"clustal"</span>,</span>
<span>  tool <span class="op">=</span> <span class="st">"pal2nal"</span>,</span>
<span>  params <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  codon_aln_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  get_aln <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>file_aln</dt>
<dd><p>a character string specifying the path to the file storing the protein alignment in CLUSTAL or FASTA format.</p></dd>


<dt>file_nuc</dt>
<dd><p>a character string specifying the path to the file storing the coding sequences in multiple FASTA format.</p></dd>


<dt>format</dt>
<dd><p>a character string specifying the file format used to store the codon alignment, e.g. "fasta", "clustal".</p></dd>


<dt>tool</dt>
<dd><p>a character string specifying the program that should be used e.g. "pal2nal".</p></dd>


<dt>params</dt>
<dd><p>a character string specifying additional parameters that shall be passed to PAL2NAL. For example
<code>params</code> = <code>"-codontable 2"</code> would specify a codon table: <em>vertebrate mitochondrial code</em>
instead if the <em>universal code</em> (default). The default value is <code>params</code> = <code>NULL</code>.
Default is <code>codon_aln_name</code> = <code>NULL</code> denoting a default name: 'tool_name.aln' .</p></dd>


<dt>codon_aln_name</dt>
<dd><p>a character string specifying the name of the stored alignment file.</p></dd>


<dt>get_aln</dt>
<dd><p>a logical value indicating whether the produced alignment should be returned.</p></dd>


<dt>quiet</dt>
<dd><p>a logical value specifying whether a successful interface call shall be printed out.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>In case <code>get_aln</code> = <code>TRUE</code> an seqinr alignment object is returned.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function provides an interface between the R language and common codon alignment tools such as "PAL2NAL".
Codon alignments can be used to quantify the evolutionary pressure acting on protein sequences based on dNdS estimation.</p>
<p>The PAL2NAL program is used to convert a multiple sequence alignment or pairwise alignment of proteins and the corresponding DNA (or mRNA) sequences into 
a codon-based DNA alignment. The output codon-based DNA alignment can then be used for <code><a href="dNdS.html">dNdS</a></code> estimation.</p>
<p>The popular codon alignment tool PAL2NAL is included in this package and is being called when choosing <code>tool</code> = <code>"pal2nal"</code>.</p>
    </div>
    <div id="note">
    <h2>Note</h2>
    <p>The PAL2NAL program automatically checks</p><ul><li><p>If you use the same IDs are used in the protein alignment file and DNA (or mRNA) file -&gt;
in this case the sequences don't have to be in the same order</p></li>
<li><p>If you don't use the same IDs are used in the protein alignment file and DNA (or mRNA) file
-&gt; you have to rearrange the IDs or sequences so that sequences in the protein alignment file and DNA (or mRNA) file
have the same order</p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>Mikita Suyama, David Torrents, and Peer Bork (2006)
PAL2NAL: robust conversion of protein sequence alignments into the corresponding codon alignments. Nucleic Acids Res. 34, W609-W612.</p>
<p>http://www.bork.embl.de/pal2nal/</p>
<p>http://www.genome.med.kyoto-u.ac.jp/cgi-bin/suyama/pal2nal/index.cgi</p>
<p>http://abacus.gene.ucl.ac.uk/software/paml.html</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="pairwise_aln.html">pairwise_aln</a></code>, <code><a href="multi_aln.html">multi_aln</a></code>, <code><a href="substitutionrate.html">substitutionrate</a></code>,
 <code><a href="dNdS.html">dNdS</a></code>, <code><a href="divergence_stratigraphy.html">divergence_stratigraphy</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Sarah Scharfenberg and Hajk-Georg Drost</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># performing a codon alignment using PAL2NAL</span></span></span>
<span class="r-in"><span><span class="va">codon_aln</span> <span class="op">&lt;-</span> <span class="fu">codon_aln</span><span class="op">(</span>file_aln <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/aa_seqs.aln'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                       file_nuc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/dna_seqs.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>                       format   <span class="op">=</span> <span class="st">"clustal"</span>, </span></span>
<span class="r-in"><span>                       tool     <span class="op">=</span> <span class="st">"pal2nal"</span>, </span></span>
<span class="r-in"><span>                       get_aln  <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span>                       </span></span>
<span class="r-in"><span>                       </span></span>
<span class="r-in"><span><span class="co"># running PAL2NAL with additional parameters (e.g. a different codon table)</span></span></span>
<span class="r-in"><span><span class="va">codon_aln</span> <span class="op">&lt;-</span> <span class="fu">codon_aln</span><span class="op">(</span>file_aln <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/aa_seqs.aln'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                       file_nuc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/dna_seqs.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>                       format   <span class="op">=</span> <span class="st">"clustal"</span>, </span></span>
<span class="r-in"><span>                       tool     <span class="op">=</span> <span class="st">"pal2nal"</span>, </span></span>
<span class="r-in"><span>                       get_aln  <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                       params   <span class="op">=</span> <span class="st">"-codontable 2"</span><span class="op">)</span></span></span>
<span class="r-in"><span>                        </span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span>                       </span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

