<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Compute dNdS Values For A Given Pairwise Alignment — compute_dnds • orthologr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Compute dNdS Values For A Given Pairwise Alignment — compute_dnds"><meta property="og:description" content="This function takes a vector containing the
query amino acid sequence, subject amino acid sequence, query CDS sequence, and subject CDS sequence
and then runs the following pipieline:

1) Multiple-Alignment of query amino acid sequence and subject amino acid sequence
2) Codon-Alignment of the amino acid alignment returned by 1) and query CDS sequence + subject CDS sequence
3) dNdS estimation of the codon alignment returned by 2)

"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">Orthology inference dNdS estimation of orthologous genes using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/drostlab/orthologr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute dNdS Values For A Given Pairwise Alignment</h1>
    <small class="dont-index">Source: <a href="https://github.com/drostlab/orthologr/blob/HEAD/R/compute_dnds.R" class="external-link"><code>R/compute_dnds.R</code></a></small>
    <div class="hidden name"><code>compute_dnds.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes a vector containing the
query amino acid sequence, subject amino acid sequence, query CDS sequence, and subject CDS sequence
and then runs the following pipieline:</p>
<ul><li><p>1) Multiple-Alignment of query amino acid sequence and subject amino acid sequence</p></li>
<li><p>2) Codon-Alignment of the amino acid alignment returned by 1) and query CDS sequence + subject CDS sequence</p></li>
<li><p>3) dNdS estimation of the codon alignment returned by 2)</p></li>
</ul></div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">compute_dnds</span><span class="op">(</span></span>
<span>  <span class="va">complete_tbl</span>,</span>
<span>  aa_aln_type <span class="op">=</span> <span class="st">"multiple"</span>,</span>
<span>  aa_aln_tool <span class="op">=</span> <span class="st">"clustalw"</span>,</span>
<span>  aa_aln_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  aa_aln_params <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  codon_aln_tool <span class="op">=</span> <span class="st">"pal2nal"</span>,</span>
<span>  dnds_est.method <span class="op">=</span> <span class="st">"YN"</span>,</span>
<span>  store_locally <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  kaks_calc_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  comp_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  clean_folders <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>complete_tbl</dt>
<dd><p>a data.table object storing the query_id, subject_id, query_cds (sequence), 
subject_cds (sequence), query_aa (sequence), and subject_aa (sequence) of the organisms that shall be compared.</p></dd>


<dt>aa_aln_type</dt>
<dd><p>a character string specifying the amino acid alignement type: <code>aa_aln_type</code> = "multiple" or <code>aa_aln_type</code> = <code>"pairwise"</code>.
Default is <code>aa_aln_type</code> = "multiple".</p></dd>


<dt>aa_aln_tool</dt>
<dd><p>a character string specifying the multiple alignment tool that shall be used for pairwise protein alignments.</p></dd>


<dt>aa_aln_path</dt>
<dd><p>a character string specifying the path to the corresponding multiple alignment tool.</p></dd>


<dt>aa_aln_params</dt>
<dd><p>a character string specifying additional parameters that shall be passed to the multiple alignment system call.</p></dd>


<dt>codon_aln_tool</dt>
<dd><p>a character string specifying the codon alignment tool that shall be used for codon alignments. Default is <code>codon_aln_tool</code> = <code>"pal2nal"</code>.</p></dd>


<dt>dnds_est.method</dt>
<dd><p>a character string specifying the dNdS estimation method, e.g. "Comeron","Li", "YN", etc. See Details for all options.</p></dd>


<dt>store_locally</dt>
<dd><p>a logical value indicating whether or not alignment files shall be stored locally rather than in <code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir()</a></code>.</p></dd>


<dt>kaks_calc_path</dt>
<dd><p>a character string specifying the execution path to KaKs_Calculator. Default is <code>kaks_calc_path</code> = <code>NULL</code>
(meaning that KaKs_Calculator is stored and executable in your default <code>PATH</code>).</p></dd>


<dt>quiet</dt>
<dd><p>a logical value specifying whether a successful interface call shall be printed out.</p></dd>


<dt>comp_cores</dt>
<dd><p>a numeric value specifying the number of cores that shall be used to perform parallel computations on a multicore machine.</p></dd>


<dt>clean_folders</dt>
<dd><p>a boolean value spefiying whether all internall folders storing the output of used programs
shall be removed. Default is <code>clean_folders</code> = <code>FALSE</code>.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>This function takes the amino acid and CDS sequences two orthologous genes
and writes the corresponding amino acid and CDS sequences as fasta file into
the internal folder environment. The resulting fasta files (two files) store the 
amino acid sequence of the query_id and subject_id (file one) and the CDS sequence of
the query_id and subject_id (file two). These fasta files are then used to pass through the following pipeline:</p>
<p>1) Multiple-Alignment or Pairwise-Alignment of query amino acid sequence and subject amino acid sequence</p>
<p>2) Codon-Alignment of the amino acid alignment returned by 1) and query CDS sequence + subject CDS sequence</p>
<p>3) dNdS estimation of the codon alignment returned by 2)</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p><a href="http://www.r-bloggers.com/the-wonders-of-foreach/" class="external-link">http://www.r-bloggers.com/the-wonders-of-foreach/</a></p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="multi_aln.html">multi_aln</a></code>, <code><a href="substitutionrate.html">substitutionrate</a></code>, <code><a href="dNdS.html">dNdS</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Hajk-Georg Drost and Sarah Scharfenberg</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

