<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ orthologr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">Orthology inference dNdS estimation of orthologous genes using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/drostlab/orthologr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="blast.html">blast()</a></code> </p>
        </td>
        <td><p>Perform a BLAST+ search</p></td>
      </tr><tr><td>
          <p><code><a href="blast_best.html">blast_best()</a></code> </p>
        </td>
        <td><p>Perform a BLAST+ best hit search</p></td>
      </tr><tr><td>
          <p><code><a href="blast_rec.html">blast_rec()</a></code> </p>
        </td>
        <td><p>Perform a BLAST+ reciprocal best hit (RBH) search</p></td>
      </tr><tr><td>
          <p><code><a href="cds2aa.html">cds2aa()</a></code> </p>
        </td>
        <td><p>Translate CDS file to Amino Acids file</p></td>
      </tr><tr><td>
          <p><code><a href="check_annotation.html">check_annotation()</a></code> </p>
        </td>
        <td><p>Check whether an annotation file contains outlier lines</p></td>
      </tr><tr><td>
          <p><code><a href="clean_all_folders.html">clean_all_folders()</a></code> </p>
        </td>
        <td><p>Delete the internal folder hierarchy</p></td>
      </tr><tr><td>
          <p><code><a href="codon_aln.html">codon_aln()</a></code> </p>
        </td>
        <td><p>Compute a Codon Alignment</p></td>
      </tr><tr><td>
          <p><code><a href="compute_dnds.html">compute_dnds()</a></code> </p>
        </td>
        <td><p>Compute dNdS Values For A Given Pairwise Alignment</p></td>
      </tr><tr><td>
          <p><code><a href="dNdS.html">dNdS()</a></code> </p>
        </td>
        <td><p>Compute dNdS values for two organisms</p></td>
      </tr><tr><td>
          <p><code><a href="diamond.html">diamond()</a></code> </p>
        </td>
        <td><p>Perform a DIAMOND2 search</p></td>
      </tr><tr><td>
          <p><code><a href="diamond_best.html">diamond_best()</a></code> </p>
        </td>
        <td><p>Perform a DIAMOND2 best hit search</p></td>
      </tr><tr><td>
          <p><code><a href="diamond_rec.html">diamond_rec()</a></code> </p>
        </td>
        <td><p>Perform a DIAMOND2 reciprocal best hit (RBH) search</p></td>
      </tr><tr><td>
          <p><code><a href="divergence_map.html">divergence_map()</a></code> </p>
        </td>
        <td><p>Sort dNdS Values Into Divergence Strata</p></td>
      </tr><tr><td>
          <p><code><a href="divergence_stratigraphy.html">divergence_stratigraphy()</a></code> </p>
        </td>
        <td><p>Perform Divergence Stratigraphy</p></td>
      </tr><tr><td>
          <p><code><a href="extract_features.html">extract_features()</a></code> </p>
        </td>
        <td><p>Helper function to extract gene loci and splice variant IDs from GFF files</p></td>
      </tr><tr><td>
          <p><code><a href="filter_best_hits.html">filter_best_hits()</a></code> </p>
        </td>
        <td><p>Helper function to select best BLAST hit based on minimum evalue</p></td>
      </tr><tr><td>
          <p><code><a href="filter_core_set.html">filter_core_set()</a></code> </p>
        </td>
        <td><p>Helper function to extract a core set of orthologous gene loci</p></td>
      </tr><tr><td>
          <p><code><a href="filter_core_set_lnc.html">filter_core_set_lnc()</a></code> </p>
        </td>
        <td><p>Helper function to extract a core set of orthologous lncRNAs</p></td>
      </tr><tr><td>
          <p><code><a href="filter_dNdS.html">filter_dNdS()</a></code> </p>
        </td>
        <td><p>Filter dNdS values</p></td>
      </tr><tr><td>
          <p><code><a href="generate_ortholog_tables.html">generate_ortholog_tables()</a></code> </p>
        </td>
        <td><p>Generate ortholog tables by gene locus and splice varaint</p></td>
      </tr><tr><td>
          <p><code><a href="generate_ortholog_tables_all.html">generate_ortholog_tables_all()</a></code> </p>
        </td>
        <td><p>Generate ortholog tables by gene locus and splice varaint for a set of species</p></td>
      </tr><tr><td>
          <p><code><a href="import_ortholog_tables_all.html">import_ortholog_tables_all()</a></code> </p>
        </td>
        <td><p>Importing output pairwise orthologs tables generated with <code>generate_ortholog_tables_all</code></p></td>
      </tr><tr><td>
          <p><code><a href="lnc_map_core_orthologs.html">lnc_map_core_orthologs()</a></code> </p>
        </td>
        <td><p>Retrieve a core set of orthologous lncRNAs from the pairwise lncRNA orthologs map</p></td>
      </tr><tr><td>
          <p><code><a href="lnc_map_counts.html">lnc_map_counts()</a></code> </p>
        </td>
        <td><p>Count number of orthologous lncRNAs per pairwise species comparison</p></td>
      </tr><tr><td>
          <p><code><a href="map_generator_dnds.html">map_generator_dnds()</a></code> </p>
        </td>
        <td><p>Generate dNdS Maps Between a Query Organism and Multiple Subject Organisms</p></td>
      </tr><tr><td>
          <p><code><a href="map_generator_lnc.html">map_generator_lnc()</a></code> </p>
        </td>
        <td><p>Infer orthologous lncRNAs between multiple species</p></td>
      </tr><tr><td>
          <p><code><a href="multi_aln.html">multi_aln()</a></code> </p>
        </td>
        <td><p>Compute Multiple Sequence Alignments</p></td>
      </tr><tr><td>
          <p><code><a href="orthofinder2.html">orthofinder2()</a></code> </p>
        </td>
        <td><p>Interface function to Orthofinder2</p></td>
      </tr><tr><td>
          <p><code><a href="orthofinder2_retrieve_core_orthologs.html">orthofinder2_retrieve_core_orthologs()</a></code> </p>
        </td>
        <td><p>Retrieve core orthologs across multiple species from Orthofinder2 output</p></td>
      </tr><tr><td>
          <p><code><a href="orthologs.html">orthologs()</a></code> </p>
        </td>
        <td><p>Main Orthology Inference Function</p></td>
      </tr><tr><td>
          <p><code><a href="orthologs_lnc.html">orthologs_lnc()</a></code> <code><a href="orthologs_lnc.html">orthologs_lnc()</a></code> </p>
        </td>
        <td><p>Orthology Inference of lncRNAs</p></td>
      </tr><tr><td>
          <p><code><a href="pairwise_aln.html">pairwise_aln()</a></code> </p>
        </td>
        <td><p>Compute Pairwise Alignments</p></td>
      </tr><tr><td>
          <p><code><a href="plot_diverse_homology_thresholds.html">plot_diverse_homology_thresholds()</a></code> </p>
        </td>
        <td><p>Diverse line plots visualizing the number of pairwise orthologs within a ortho table generated with <code>generate_ortholog_tables_all</code>
based on different sets of homology thresholds.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_diverse_homology_thresholds_core_orthologs.html">plot_diverse_homology_thresholds_core_orthologs()</a></code> </p>
        </td>
        <td><p>Diverse line plots visualizing the number of core orthologs within a ortho table generated with <code>generate_ortholog_tables_all</code>
based on different sets of homology thresholds.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_pairwise_orthologs.html">plot_pairwise_orthologs()</a></code> </p>
        </td>
        <td><p>A line plot visualizing the number of pairwise orthologs within a ortho table generated with <code>generate_ortholog_tables_all</code></p></td>
      </tr><tr><td>
          <p><code><a href="promotor_divergence_estimation.html">promotor_divergence_estimation()</a></code> </p>
        </td>
        <td><p>Estimate the DNA distance between promotor sequences</p></td>
      </tr><tr><td>
          <p><code><a href="promotor_divergence_of_orthologous_genes.html">promotor_divergence_of_orthologous_genes()</a></code> </p>
        </td>
        <td><p>Compute promotor sequence divergence of orthologous genes</p></td>
      </tr><tr><td>
          <p><code><a href="read.cds.html">read.cds()</a></code> </p>
        </td>
        <td><p>Read the CDS of a given organism</p></td>
      </tr><tr><td>
          <p><code><a href="read.dnds.tbl.html">read.dnds.tbl()</a></code> </p>
        </td>
        <td><p>Import a dNdS table generated with <code>dNdS</code></p></td>
      </tr><tr><td>
          <p><code><a href="read.genome.html">read.genome()</a></code> </p>
        </td>
        <td><p>Read the genome of a given organism</p></td>
      </tr><tr><td>
          <p><code><a href="read.proteome.html">read.proteome()</a></code> </p>
        </td>
        <td><p>Read the proteome of a given organism</p></td>
      </tr><tr><td>
          <p><code><a href="retrieve_core_orthologs.html">retrieve_core_orthologs()</a></code> </p>
        </td>
        <td><p>Retrieve a core set of orthologous gene loci from several pairwise ortholog tables</p></td>
      </tr><tr><td>
          <p><code><a href="retrieve_longest_isoforms.html">retrieve_longest_isoforms()</a></code> </p>
        </td>
        <td><p>Retrieve the longest isoforms from a proteome file and save results as fasta file</p></td>
      </tr><tr><td>
          <p><code><a href="retrieve_longest_isoforms_all.html">retrieve_longest_isoforms_all()</a></code> </p>
        </td>
        <td><p>Retrieve the longest isoforms of several proteome files stored in a folder</p></td>
      </tr><tr><td>
          <p><code><a href="select_orthologs.html">select_orthologs()</a></code> </p>
        </td>
        <td><p>Select orthologs based on either gene locus or splice variant</p></td>
      </tr><tr><td>
          <p><code><a href="set_blast.html">set_blast()</a></code> </p>
        </td>
        <td><p>Create a BLASTable database with makeblastdb</p></td>
      </tr><tr><td>
          <p><code><a href="set_diamond.html">set_diamond()</a></code> </p>
        </td>
        <td><p>Create a DIAMONDable database with <code>diamond makedb</code></p></td>
      </tr><tr><td>
          <p><code><a href="substitutionrate.html">substitutionrate()</a></code> </p>
        </td>
        <td><p>Internal function for dNdS computations</p></td>
      </tr><tr><td>
          <p><code><a href="testCoreOrthoParamsGeneLocus.html">testCoreOrthoParamsGeneLocus()</a></code> </p>
        </td>
        <td><p>Helper function for <code>plot_diverse_homology_thresholds_core_orthologs</code></p></td>
      </tr><tr><td>
          <p><code><a href="testCoreOrthoParamsSpliceVariant.html">testCoreOrthoParamsSpliceVariant()</a></code> </p>
        </td>
        <td><p>Helper function for splice variant based <code>plot_diverse_homology_thresholds_core_orthologs</code></p></td>
      </tr><tr><td>
          <p><code><a href="transl.html">transl()</a></code> </p>
        </td>
        <td><p>Translate DNA to Amino Acids</p></td>
      </tr><tr><td>
          <p><code><a href="translate_cds_to_protein.html">translate_cds_to_protein()</a></code> </p>
        </td>
        <td><p>Translate coding sequences into amino acid sequences</p></td>
      </tr><tr><td>
          <p><code><a href="translate_cds_to_protein_all.html">translate_cds_to_protein_all()</a></code> </p>
        </td>
        <td><p>Translate coding sequences into amino acid sequences for multiple files</p></td>
      </tr><tr><td>
          <p><code><a href="write.proteome.html">write.proteome()</a></code> </p>
        </td>
        <td><p>Save a proteome in fasta format</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

