<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate dNdS Maps Between a Query Organism and Multiple Subject Organisms — map_generator_dnds • orthologr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Generate dNdS Maps Between a Query Organism and Multiple Subject Organisms — map_generator_dnds"><meta property="og:description" content="This function allows you to compute dNdS or DS Maps between a query organism
and a set subject organisms stored in the same folder. The corresponding dNdS/DS Maps are then stored in an output folder."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">Orthology inference dNdS estimation of orthologous genes using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/drostlab/orthologr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate dNdS Maps Between a Query Organism and Multiple Subject Organisms</h1>
    <small class="dont-index">Source: <a href="https://github.com/drostlab/orthologr/blob/HEAD/R/map.generator.R" class="external-link"><code>R/map.generator.R</code></a></small>
    <div class="hidden name"><code>map_generator_dnds.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function allows you to compute dNdS or DS Maps between a query organism
and a set subject organisms stored in the same folder. The corresponding dNdS/DS Maps are then stored in an output folder.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">map_generator_dnds</span><span class="op">(</span></span>
<span>  <span class="va">query_file</span>,</span>
<span>  <span class="va">subjects_folder</span>,</span>
<span>  <span class="va">output_folder</span>,</span>
<span>  eval <span class="op">=</span> <span class="st">"1E-5"</span>,</span>
<span>  min_qry_coverage_hsp <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  min_qry_perc_identity <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  ortho_detection <span class="op">=</span> <span class="st">"RBH"</span>,</span>
<span>  aa_aln_type <span class="op">=</span> <span class="st">"pairwise"</span>,</span>
<span>  aa_aln_tool <span class="op">=</span> <span class="st">"NW"</span>,</span>
<span>  codon_aln_tool <span class="op">=</span> <span class="st">"pal2nal"</span>,</span>
<span>  dnds_est_method <span class="op">=</span> <span class="st">"Comeron"</span>,</span>
<span>  comp_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  progress_bar <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">";"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>query_file</dt>
<dd><p>a character string specifying the path to the CDS file of the query organism.</p></dd>


<dt>subjects_folder</dt>
<dd><p>a character string specifying the path to the folder where CDS files of the subject organisms are stored.</p></dd>


<dt>output_folder</dt>
<dd><p>a character string specifying the path to the folder where output dNdS/DS Maps should be stored.</p></dd>


<dt>eval</dt>
<dd><p>a character string specifying the e-value for BLAST based Orthology Inference that is performed
in the process of dNdS computations. Please use the scientific notation.</p></dd>


<dt>min_qry_coverage_hsp</dt>
<dd><p>minimum <code>qcovhsp</code> (= query coverage of the HSP) of an orthologous hit (a value between 1 and 100).</p></dd>


<dt>min_qry_perc_identity</dt>
<dd><p>minimum <code>perc_identity</code> (= percent sequence identity between query and selected HSP) of an orthologous hit (a value between 1 and 100).</p></dd>


<dt>ortho_detection</dt>
<dd><p>a character string specifying the Orthology Inference method that shall be used to perform
dNdS computations. Possible options are: <code>ortho_detection</code> = <code>"BH"</code> (BLAST best hit), 
<code>ortho_detection</code> = <code>"RBH"</code> (BLAST best reciprocal hit), etc.</p></dd>


<dt>aa_aln_type</dt>
<dd><p>a character string specifying the amino acid alignement type: <code>aa_aln_type = "multiple"</code> or <code>aa_aln_type = "pairwise"</code>. 
Default is <code>aa_aln_type = "pairwise"</code>.</p></dd>


<dt>aa_aln_tool</dt>
<dd><p>a character string specifying the program that should be used e.g. "clustalw".</p></dd>


<dt>codon_aln_tool</dt>
<dd><p>a character string specifying the codon alignment tool that shall be used. 
Default is <code>codon_aln_tool = "pal2nal"</code>. Right now only "pal2nal" can be selected as codon alignment tool.</p></dd>


<dt>dnds_est_method</dt>
<dd><p>a character string specifying the dNdS estimation method, e.g. "Comeron","Li", "YN", etc. See Details for all options.</p></dd>


<dt>comp_cores</dt>
<dd><p>number of computing cores that shall be used to perform parallelized computations.</p></dd>


<dt>progress_bar</dt>
<dd><p>should a progress bar be shown. Default is <code>progress_bar = TRUE</code>.</p></dd>


<dt>sep</dt>
<dd><p>a file separator that is used to store maps as csv file.</p></dd>


<dt>...</dt>
<dd><p>additional parameters that shall be passed to  <code><a href="dNdS.html">dNdS</a></code>.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Given a query organism and a set of subject organsisms that are stored in the same folder,
this function crawls through all subject organsism and as a first step computes the pairwise 
dNdS Maps between query and subject organsim and as a second step stores the corresponding Map 
in an output folder.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Hajk-Georg Drost</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="fu">map_generator_dnds</span><span class="op">(</span></span></span>
<span class="r-in"><span>   query_file      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_thal_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   subjects_folder <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/map_gen_example'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   aa_aln_type      <span class="op">=</span> <span class="st">"pairwise"</span>,</span></span>
<span class="r-in"><span>   aa_aln_tool      <span class="op">=</span> <span class="st">"NW"</span>, </span></span>
<span class="r-in"><span>   codon_aln_tool   <span class="op">=</span> <span class="st">"pal2nal"</span>, </span></span>
<span class="r-in"><span>   dnds_est_method  <span class="op">=</span> <span class="st">"Comeron"</span>,</span></span>
<span class="r-in"><span>   output_folder   <span class="op">=</span> <span class="st">"orthologr_dnds_maps"</span>,</span></span>
<span class="r-in"><span>   quiet           <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>   comp_cores      <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

