<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Infer orthologous lncRNAs between multiple species — map_generator_lnc • orthologr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Infer orthologous lncRNAs between multiple species — map_generator_lnc"><meta property="og:description" content="Inference of orthologous lncRNAs between multiple species is performed via pairwise BLAST (reciprocal) best hit comparisons.
 The corresponding orthologous tables are then stored in an output folder."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">Orthology inference dNdS estimation of orthologous genes using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/drostlab/orthologr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Infer orthologous lncRNAs between multiple species</h1>
    <small class="dont-index">Source: <a href="https://github.com/drostlab/orthologr/blob/HEAD/R/map.generator.lnc.R" class="external-link"><code>R/map.generator.lnc.R</code></a></small>
    <div class="hidden name"><code>map_generator_lnc.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Inference of orthologous lncRNAs between multiple species is performed via pairwise BLAST (reciprocal) best hit comparisons.
 The corresponding orthologous tables are then stored in an output folder.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">map_generator_lnc</span><span class="op">(</span></span>
<span>  <span class="va">query_file</span>,</span>
<span>  <span class="va">subjects_folder</span>,</span>
<span>  <span class="va">output_folder</span>,</span>
<span>  task <span class="op">=</span> <span class="st">"blastn"</span>,</span>
<span>  eval <span class="op">=</span> <span class="st">"1E-5"</span>,</span>
<span>  ortho_detection <span class="op">=</span> <span class="st">"RBH"</span>,</span>
<span>  max.target.seqs <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  min_qry_coverage_hsp <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  min_qry_perc_identity <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  logical_connective <span class="op">=</span> <span class="st">"AND"</span>,</span>
<span>  min_alig_length <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  comp_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  progress_bar <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">";"</span>,</span>
<span>  path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>query_file</dt>
<dd><p>a character string specifying the path to the lncRNAs file of the query organism in <code>fasta</code> format.</p></dd>


<dt>subjects_folder</dt>
<dd><p>a character string specifying the path to the folder where lncRNAs files in <code>fasta</code> format of the subject organisms are stored.</p></dd>


<dt>output_folder</dt>
<dd><p>a character string specifying the path to the folder where output orthologous tables should be stored.</p></dd>


<dt>task</dt>
<dd><p>nucleotide search task option. Options are:</p><ul><li><p><code>task = "blastn"</code> : Standard nucleotide-nucleotide comparisons (default) - Traditional BLASTN requiring an exact match of 11.</p></li>
<li><p><code>task = "blastn-short"</code> : Optimized nucleotide-nucleotide comparisons for query sequences shorter than 50 nucleotides.</p></li>
<li><p><code>task = "dc-megablast"</code> : Discontiguous megablast used to find somewhat distant sequences.</p></li>
<li><p><code>task = "megablast"</code> : Traditional megablast used to find very similar (e.g., intraspecies or closely related species) sequences.</p></li>
<li><p><code>task = "rmblastn"</code></p></li>
</ul></dd>


<dt>eval</dt>
<dd><p>a character string specifying the e-value for BLAST based orthology inference. Please use the scientific notation.</p></dd>


<dt>ortho_detection</dt>
<dd><p>a character string specifying the Orthology Inference method that shall be used to perform
dNdS computations. Possible options are:</p><ul><li><p><code>ortho_detection = "BH"</code>: BLAST best unidirectional hit</p></li>
<li><p><code>ortho_detection = "RBH"</code>: BLAST best reciprocal hit</p></li>
</ul></dd>


<dt>max.target.seqs</dt>
<dd><p>a numeric value specifying the number of aligned sequences to keep.
Please be aware that <code>max.target.seqs</code> selects best hits based on the database entry and not by the best e-value. See details here: https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/bty833/5106166 .</p></dd>


<dt>min_qry_coverage_hsp</dt>
<dd><p>minimum <code>qcovhsp</code> (= query coverage of the HSP) of an orthologous hit (a value between 1 and 100).</p></dd>


<dt>min_qry_perc_identity</dt>
<dd><p>minimum <code>perc_identity</code> (= percent sequence identity between query and selected HSP) of an orthologous hit (a value between 1 and 100).</p></dd>


<dt>logical_connective</dt>
<dd><p>character representing logical connective (either "AND" or "OR") if <code>min_alig_length</code> is not <code>NULL</code> filtering is done on <code>min_alig_length</code> and/or <code>min_qry_perc_identity</code></p></dd>


<dt>min_alig_length</dt>
<dd><p>minimum <code>alig_length</code> (alignment length) to an orthologous hit (number of aligned nucleotides or amino acids depending on the input data)</p></dd>


<dt>comp_cores</dt>
<dd><p>number of computing cores that shall be used to perform parallelized computations.</p></dd>


<dt>progress_bar</dt>
<dd><p>should a progress bar be shown. Default is <code>progress_bar = TRUE</code>.</p></dd>


<dt>sep</dt>
<dd><p>a file separator that is used to store maps as csv file.</p></dd>


<dt>path</dt>
<dd><p>a character string specifying the path to the corresponding orthology inference tool.
For "BH" and "RBH": path to BLAST, "PO": path to ProteinOrtho 5.07, "OrthoMCL": path to OrthoMCL.</p></dd>


<dt>...</dt>
<dd><p>additional parameters that shall be passed to  <code><a href="dNdS.html">dNdS</a></code>.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Given a query organism and a set of subject organsisms that are stored in the same folder,
this function crawls through all subject organsism and infers the lncRNA homologs in
pairwise species comparisons.</p>
    </div>
    <div id="note">
    <h2>Note</h2>
    <p>According to Sarropoulos, I., et al. (2019) orthology detection of lncRNAs was performed by reciprocal BLAST searches. Significant hits with an e-value &lt;= 10-3 were selected having an alignment identity &gt;= 10% OR a minimum alignment length &gt;= 50 nucleotides.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Sarropoulos I, Marin R, Cardoso-Moreira M, Kaessmann H (2019).
“Developmental dynamics of lncRNAs across mammalian organs and species.”
<em>Nature</em>, <b>571</b>, 510--514.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Hajk-Georg Drost</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># example using classic blastn searches</span></span></span>
<span class="r-in"><span><span class="fu">map_generator_lnc</span><span class="op">(</span></span></span>
<span class="r-in"><span>   query_file      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_thal_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   subjects_folder <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/map_gen_example'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   output_folder   <span class="op">=</span> <span class="st">"orthologs_lncrna"</span>,</span></span>
<span class="r-in"><span>   comp_cores      <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># example using  discontiguous megablast used to find somewhat distant sequences</span></span></span>
<span class="r-in"><span><span class="fu">map_generator_lnc</span><span class="op">(</span></span></span>
<span class="r-in"><span>   query_file      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_thal_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   subjects_folder <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/map_gen_example'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   output_folder   <span class="op">=</span> <span class="st">"orthologs_lncrna"</span>,</span></span>
<span class="r-in"><span>   task            <span class="op">=</span> <span class="st">"dc-megablast"</span>,</span></span>
<span class="r-in"><span>   comp_cores      <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># parameter settings based on Sarropoulos, I., et al. (2019)</span></span></span>
<span class="r-in"><span><span class="fu">map_generator_lnc</span><span class="op">(</span></span></span>
<span class="r-in"><span>   <span class="va">query_file</span>,,</span></span>
<span class="r-in"><span>   <span class="va">subjects_folder</span>,</span></span>
<span class="r-in"><span>   eval                  <span class="op">=</span> <span class="fl">1E-3</span>,</span></span>
<span class="r-in"><span>   ortho_detection       <span class="op">=</span> <span class="st">"RBH"</span>,</span></span>
<span class="r-in"><span>   <span class="va">output_folder</span>,</span></span>
<span class="r-in"><span>   min_qry_coverage_hsp  <span class="op">=</span> <span class="fl">0</span>,</span></span>
<span class="r-in"><span>   min_qry_perc_identity <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>   logical_connective    <span class="op">=</span> <span class="st">"OR"</span>,</span></span>
<span class="r-in"><span>   min_alig_length       <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

