<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Interface function to Orthofinder2 — orthofinder2 • orthologr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Interface function to Orthofinder2 — orthofinder2"><meta property="og:description" content="Run Orthofinder2 from R."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">Orthology inference dNdS estimation of orthologous genes using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/drostlab/orthologr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Interface function to Orthofinder2</h1>
    <small class="dont-index">Source: <a href="https://github.com/drostlab/orthologr/blob/HEAD/R/orthofinder2.R" class="external-link"><code>R/orthofinder2.R</code></a></small>
    <div class="hidden name"><code>orthofinder2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run Orthofinder2 from R.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">orthofinder2</span><span class="op">(</span></span>
<span>  <span class="va">proteome_folder</span>,</span>
<span>  use_existing_output <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  import_type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  comp_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  of_path <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>proteome_folder</dt>
<dd><p>file path to a folder storing the proteome sequences of the species for which orthology inference shall be performed.</p></dd>


<dt>use_existing_output</dt>
<dd><p>a logical value indicating whether or not an existing <code>Orthofinder2</code>
output folder shall be used fo further import and processing. If <code>use_existing_output = TRUE</code>
is selected then please specify the file path to the in <code>proteome_folder</code> where the <code>Orthofinder2</code> output folder can be found.</p></dd>


<dt>import_type</dt>
<dd><p>type of <code>Orthofinder2</code> output that shall be imported after running <code>Orthofinder2</code>.
Options are:</p><ul><li><p><code>import_type = "orthogroups_core"</code></p></li>
<li><p><code>import_type = "orthogroups_pairwise"</code></p></li>
<li><p><code>import_type = ""</code></p></li>
<li><p><code>import_type = ""</code></p></li>
</ul></dd>


<dt>comp_cores</dt>
<dd><p>number of cores that shall be used for parallel processing. Default is <code>cores = 1</code>.</p></dd>


<dt>of_path</dt>
<dd><p>a character string specifying the path to the locally installed <code>orthofinder</code> executable.
A possible specification could be <code>of_path = "/opt/miniconda3/bin/"</code> which internally will translate to
<code>/opt/miniconda3/bin/orthofinder</code>. The default is <code>of_path = NULL</code> which means that orthofinder assumes
users have their <code>orthofinder</code> executable stored at <code>/opt/miniconda3/bin/orthofinder</code>.</p></dd>

</dl></div>
    <div id="note">
    <h2>Note</h2>
    <p>This function assumes that users have <code>OrthoFinder</code> installed via <code>miniconda</code> and stored at <code>~/opt/miniconda3/bin/</code>.
In addition, DIAMOND needs to be installed as executable tool (/usr/local/bin).</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Hajk-Georg Drost</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># specify species names</span></span></span>
<span class="r-in"><span><span class="va">orgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Arabidopsis lyrata"</span>,</span></span>
<span class="r-in"><span>          <span class="st">"Capsella rubella"</span>, <span class="st">"Solanum lycopersicum"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># download proteome files for all species</span></span></span>
<span class="r-in"><span><span class="fu">biomartr</span><span class="fu">::</span><span class="fu">getProteomeSet</span><span class="op">(</span>db <span class="op">=</span> <span class="st">"refseq"</span>, organisms <span class="op">=</span> <span class="va">orgs</span>, path <span class="op">=</span> <span class="st">"of_proteomes"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># download annotation files for all species</span></span></span>
<span class="r-in"><span><span class="fu">biomartr</span><span class="fu">::</span><span class="fu">getGFFSet</span><span class="op">(</span>db <span class="op">=</span> <span class="st">"refseq"</span>, organisms <span class="op">=</span> <span class="va">orgs</span>, path <span class="op">=</span> <span class="st">"of_gff"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># select longest splice variant per gene locus</span></span></span>
<span class="r-in"><span><span class="fu"><a href="retrieve_longest_isoforms_all.html">retrieve_longest_isoforms_all</a></span><span class="op">(</span>proteome_folder <span class="op">=</span> <span class="st">"of_proteomes"</span>,</span></span>
<span class="r-in"><span>                              annotation_folder <span class="op">=</span> <span class="st">"of_gff"</span>,</span></span>
<span class="r-in"><span>                              annotation_format <span class="op">=</span> <span class="st">"gff"</span>,</span></span>
<span class="r-in"><span>                              output_folder <span class="op">=</span> <span class="st">"of_proteomes_longest_sv"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># run orthofinder2 to infer ortho groups for the specified species</span></span></span>
<span class="r-in"><span><span class="fu">orthofinder2</span><span class="op">(</span>proteome_folder <span class="op">=</span> <span class="st">"of_proteomes_longest_sv"</span>, comp_cores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

