<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create a DIAMONDable database with diamond makedb — set_diamond • orthologr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Create a DIAMONDable database with diamond makedb — set_diamond"><meta property="og:description" content='This function reads a file storing a specific sequence type, such as "cds", "protein", or
"dna" in a standard sequence file format such as "fasta", etc. and depending of the makedb
 parameter either creates a diamond-able database, or returns the corresponding protein sequences
 as data.table object for further DIAMOND2 searches.'><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">Orthology inference dNdS estimation of orthologous genes using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/drostlab/orthologr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a DIAMONDable database with <code>diamond makedb</code></h1>
    <small class="dont-index">Source: <a href="https://github.com/drostlab/orthologr/blob/HEAD/R/set_diamond.R" class="external-link"><code>R/set_diamond.R</code></a></small>
    <div class="hidden name"><code>set_diamond.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function reads a file storing a specific sequence type, such as "cds", "protein", or
"dna" in a standard sequence file format such as "fasta", etc. and depending of the <code>makedb</code>
 parameter either creates a diamond-able database, or returns the corresponding protein sequences
 as data.table object for further DIAMOND2 searches.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">set_diamond</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  seq_type <span class="op">=</span> <span class="st">"cds"</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"fasta"</span>,</span>
<span>  makedb <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  delete_corrupt_cds <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  makedb_type <span class="op">=</span> <span class="st">"protein"</span>,</span>
<span>  comp_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>file</dt>
<dd><p>a character string specifying the path to the file storing the sequences of interest.</p></dd>


<dt>seq_type</dt>
<dd><p>a character string specifying the sequence type stored in the input file.
Options are are: "cds", "protein", or "dna". In case of "cds", sequence are translated to protein sequences,
in case of "dna", cds prediction is performed on the corresponding sequences which subsequently are
translated to protein sequences. Default is <code>seq_type</code> = "cds".</p></dd>


<dt>format</dt>
<dd><p>a character string specifying the file format used to store the genome, e.g. "fasta", "gbk".</p></dd>


<dt>makedb</dt>
<dd><p>TRUE or FALSE whether a database should be created or not (<code>diamond makedb</code>).</p></dd>


<dt>delete_corrupt_cds</dt>
<dd><p>a logical value indicating whether sequences with corrupt base triplets should be removed from the input <code>file</code>. This is the case when the length of coding sequences cannot be divided by 3 and thus the coding sequence contains at least one corrupt base triplet.</p></dd>


<dt>path</dt>
<dd><p>a character string specifying the path to the DIAMOND2 program (in case you don't use the default path).</p></dd>


<dt>makedb_type</dt>
<dd><p>a character string specifying the sequence type stored in the DIAMOND2 database
that is generated using 'diamond makedb'. Currently, the only option is "protein". Default is <code>makedb_type</code> = "protein".</p></dd>


<dt>comp_cores</dt>
<dd><p>a numeric value specifying the number of cores to be used for multicore 'diamond makedb' computations.</p></dd>


<dt>quiet</dt>
<dd><p>a logical value indicating whether <code>diamond makedb</code> should be run with the quiet mode.
Default is <code>quiet</code> = <code>TRUE</code> (which adds <code>--quiet</code> to the diamond makedb run).</p></dd>


<dt>...</dt>
<dd><p>additional arguments that are used by the seqinr::read.fasta() function.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list storing two elements. The first element [[1]] corresponds to the data.table storing the gene ids in the first column and
the corresponding dna (cds) sequence in the second column and the aminoacid sequence third column.
The second list element [[2]] stores the name of the protein database that was created by 'diamond makedb'.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Buchfink, B., Reuter, K., &amp; Drost, H. G. (2021) "Sensitive protein alignments at tree-of-life scale using DIAMOND." Nature methods, 18(4), 366-368.</p>
<p>https://github.com/bbuchfink/diamond/wiki/3.-Command-line-options</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="diamond_best.html">diamond_best</a></code>, <code><a href="diamond_rec.html">diamond_rec</a></code>, <code><a href="diamond.html">diamond</a></code>, <code><a href="set_blast.html">set_blast</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Jaruwatana Sodai Lotharukpong</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span> <span class="co"># running the set function to see an example output</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">set_diamond</span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/ortho_thal_cds.fasta'</span>, package <span class="op">=</span> <span class="st">'orthologr'</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> , <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

